### DISCUSSION TITLE
Volumes and large point cloud interpolation

### DISCUSSION THREAD

**Discussion by gbene** [pyvista/pyvista#2752] [Q&A]
Hello,
I'm currently trying to create a 3D GPR data volume using Uniform grids. Since the dataset is pretty big about 338,310,000 points I was wondering if there are any resources or more effective methods to create the volume other than interpolate the point cloud using grid.interpolate. For example how were the knee_full or brain example volumes constructed? 

Thanks!

**Comment by gbene** [comment]
Hello,

I found a way around it but still needs some work. I define a regular grid of points and interpolate at different depths using scipy.interpolate.griddata. The interpolated values are then assigned to the cells a UniformGrid. The problem of this process is that the resulting mesh is flipped and I can't seem to find a way to assign the PC point data to the proper UniformGrid cell. I will illustrate the code and the different steps taken.

The intensity data are written in SEGY files organized in segments. Every segment is read and the different properties appendend in the appropriate list

```python

files = glob.glob(path/to/data.sgy)

# print(files)

list_n_traces = []
list_n_samples = []
list_time_range = []
list_depth = []
list_scalar = []
list_header_data = []
list_data_stream = []
list_segments = []
list_levels = []

for file in files: #for every trace

    n_traces,n_samples,time_range,depth,scalar,header_data,data_stream = SEGY_PRISM_header(file,use_scal=False)
    list_n_traces.append(n_traces)
    list_n_samples.append(n_samples)
    list_time_range.append(time_range)
    list_depth.append(depth)
    list_scalar.append(scalar)
    list_header_data.append(header_data)
    list_data_stream.append(np.array(data_stream))
    list_segments.append(header_data.loc[:,['lon','lat','alt']].values)

    
markers = [header_data.loc[header_data['mark']>0,['lon','lat','alt']].values for header_data in list_header_data]


```

The important lists here are list_data_stream, list_segments and markers. The first is a list of the trace intensity values for each point of the given segment. The second is a list of coordinates for each trace point. The third is a list of points of interest marked during the acquisition.

With list_segments we can preview the acquisition geometry and markers 

```python
plotter=pv.Plotter()

for i,segment in enumerate(list_segments):
        seg = pv.Spline(segment,20)
        seg.cell_data['segment number'] = i+1
        plotter.add_mesh(seg)

for marker in markers:
    marker = pv.PolyData(marker)
    plotter.add_mesh(marker,'r',point_size=10)
plotter.add_camera_orientation_widget()
plotter.show()
```
![image](https://user-images.githubusercontent.com/40857804/174435388-7e0372cb-c40c-4a2f-b9c0-794856fd7a33.png)

In the acquisition geometry we can see a hole that is delimited by the markers. This is fundamental because we can use this "feature" as a reference.

After doing some processing and filtering for every trace we compile a points and data array that respectively contain the position information and the segment data. 

```python
n_samples = 1024 #samples per trace
points = np.array([]).reshape(-1,3)


for pos in list_segments:
    points = np.append(points,pos,axis=0)

data_arr = np.zeros((points.shape[0],n_samples))

i=0
for data in list_data_stream:
    data_arr[i:i+data.shape[0],:data.shape[1]] = data
    i+=data.shape[0]

print(points.shape)
print(data_arr.shape)

> (26265, 3)
> (26265, 1024)
```

For the data_array we need to fill a np.zeros array of dimensions (total_points, n_samples_per_trace). This is because different segments can have different amounts of points (traces) and during the processing step the number of samples can be less than the original number and different from trace to trace.

The next step is to define a maximum depth to reach and the integration step (number of slices). We then define the boundaries and resolution of our grid  

```python

tot_depth = 2.205 # maximum depth reached by the signal during the acquisition 

conv = n_samples/tot_depth

max_depth = 1 # maximum depth to reach in the interpolation

max_depthi = int(max_depth*conv) # max_depth value in sample number


int_step = 0.1 # integration step in m. 1/int_step is the number of slices (in this case 100)
int_stepi = int(int_step*conv) # int_step value in sample number

x_min = points[:,0].min()
x_max = points[:,0].max()
y_min = points[:,1].min()
y_max = points[:,1].max()


x_range = x_max-x_min
y_range = y_max-y_min
z_range = z_max-z_min

x_res = 600
y_res = 600
z_res = int(1/int_step)

ugrid = pv.UniformGrid()

ugrid.origin = (x_min,y_min,-max_depth)
ugrid.dimensions = (x_res,y_res,z_res+1)
ugrid.spacing = (x_range/(x_res-1),y_range/(y_res-1),max_depth/(z_res))

```

We can visualize a rough (10,10,1) Uniform grid structure with the markers.

```python
max_depth = 1 # maximum depth to reach in the interpolation

max_depthi = int(max_depth*conv) # max_depth value in sample number


int_step = 1 # integration step in m. 1/int_step is the number of slices (in this case 100)
int_stepi = int(int_step*conv) # int_step value in sample number

x_min = points[:,0].min()
x_max = points[:,0].max()
y_min = points[:,1].min()
y_max = points[:,1].max()


x_range = x_max-x_min
y_range = y_max-y_min
z_range = z_max-z_min

x_res = 10
y_res = 10
z_res = int(1/int_step)

ugrid = pv.UniformGrid()

ugrid.origin = (x_min,y_min,-max_depth)
ugrid.dimensions = (x_res,y_res,z_res+1)
ugrid.spacing = (x_range/(x_res-1),y_range/(y_res-1),max_depth/(z_res))

plotter = pv.Plotter(notebook=False)
plotter.add_mesh(ugrid,style='wireframe')

for marker in markers:
    marker = pv.PolyData(marker)
    plotter.add_mesh(marker,'r',point_size=10)
plotter.add_camera_orientation_widget()
plotter.show()

```

![image](https://user-images.githubusercontent.com/40857804/174435373-07e6c71a-d379-43ea-9e55-2cae9e238ab6.png)

Then we define the regular grid and where to interpolate at each level. Each level is then appended to create a final PC composed of each level. 

```python

xrng = np.linspace(x_min, x_max, x_res, dtype=np.float32)

yrng = np.linspace(y_min, y_max, y_res, dtype=np.float32)

levels = np.linspace(0,max_depth,z_res)

for c in np.flip(levels): # The array is flipped because UniformGrid starts at -max_depth
    start = int(c*conv)
    end = int(start+int_stepi)
    data_slice = test_arr[:,start:end].mean(1) # We use a mean value of the intensities in the given depth range 

    x, y, z = np.meshgrid(xrng, yrng, -c)

    grid = pv.StructuredGrid(x, y, z)

    centers = grid.cell_centers().points # we define the position of the interpolated value as the center of the given pixel of the regular grid. 
    
    int_data = griddata(points[:,:2],data_slice,(centers[:,0],centers[:,1]),method='linear') # Interpolate the data
    
    pc_centers = pv.PolyData(centers) # Pointcloud
    
    pc_centers['intensity'] = int_data #assign the interpolated values to the points in the PointCloud
    
    appender.AddInputData(pc_centers) # Append each level

appender.Update()
pc.ShallowCopy(appender.GetOutput()) # Create final point cloud
ugrid['intensity'] = pc['intensity']
```
Finally we can plot the results with the markers


```python
plotter = pv.Plotter()
plotter.add_mesh(ugrid)

for marker in markers:
    marker = pv.PolyData(marker)
    plotter.add_mesh(marker,'r')
plotter.add_camera_orientation_widget()
plotter.show()
```
![image](https://user-images.githubusercontent.com/40857804/174435605-4c66f980-779d-4373-8ece-c28c9e0673fd.png)

The problem is in this last step. When we view the volume the information is misplaced (the "hole" is in another position). 

This is how it should be oriented. Here I visualize the pc slice at 20cm below the surface with the markers

```python
plotter = pv.Plotter()

for c in np.flip(levels)[-2:-1]: 
    start = int(c*conv)
    end = int(start+int_stepi)
    data_slice = data_arr[:,start:end].mean(1) # We use a mean value of the intensities in the given depth range 

    x, y, z = np.meshgrid(xrng, yrng, -c)

    grid = pv.StructuredGrid(x, y, z)

    centers = grid.cell_centers().points # we define the position of the interpolated value as the center of the given pixel of the regular grid. 
    
    int_data = griddata(points[:,:2],data_slice,(centers[:,0],centers[:,1]),method='linear') # Interpolate the data
    
    pc_centers = pv.PolyData(centers) # Pointcloud
    
    pc_centers['intensity'] = int_data #assign the interpolated values to the points in the PointCloud

    plotter.add_mesh(pc_centers)


for marker in markers:
    marker = pv.PolyData(marker)
    plotter.add_mesh(marker,'r')
plotter.add_camera_orientation_widget()
plotter.show()
```
![image](https://user-images.githubusercontent.com/40857804/174435325-87c6da3e-0a98-4f41-96af-0149c5f91962.png)



How can I fix this? I think that the issue lies in how the PC values are assigned to the UniformGrid cells and I can't figure out how to specify that the points in the Point cloud are ordered in slices orthogonal to the Z axis and not to X or Y. 

Thanks!

**Comment by gbene** [comment]
Ok I found the problem.

By setting ugrid['intensity'] = pc_centers['intensity'] the indexes between the two object are switched. ugrid cells are ordered along X while the pc along the Y. To resolve this I search the index of the cell containing a given point with _find_containing_cell_ and assign to that cell the associated point value. Here an example

```python
x_min = 0
x_max = 5
y_min = 0
y_max = 5

x_range = x_max-x_min
y_range = y_max-y_min
ratio = y_range/x_range

x_res = 6
y_res = 6
z_res = 1

xrng = np.linspace(x_min, x_max, x_res, dtype=np.float32)

yrng = np.linspace(y_min, y_max, y_res, dtype=np.float32)

x, y, z = np.meshgrid(xrng, yrng, 0)

grid = pv.StructuredGrid(x, y, z)

    
centers = grid.cell_centers().points

centers_pc = pv.PolyData(centers)
# Assign a value to the points. In this case is the order (first point = 1, second = 2 etctc) 
# in the real case is the interpolated value
order_values = [i for i in range(len(centers))]

centers_pc['order'] = order_values

correct_ugrid = pv.UniformGrid()

correct_ugrid.origin = (x_min,y_min,0)
correct_ugrid.dimensions = (x_res,y_res,z_res+1)
correct_ugrid.spacing = (x_range/(x_res-1),y_range/(y_res-1),1/(z_res))

plotter = pv.Plotter(notebook=False)

plotter.add_mesh(centers_pc)
plotter.add_mesh(grid,style = 'wireframe')
plotter.add_camera_orientation_widget()
plotter.show()
```
![image](https://user-images.githubusercontent.com/40857804/174856249-6ab2f1a8-afa7-4d87-83e4-57c9164cc548.png)


The correct uniform grid is obtained by:
```python
correct_ugrid['order'] = [0]*correct_ugrid.n_cells # init empty data 
indexes = correct_ugrid.find_containing_cell(centers_pc.points)
for i,ind in enumerate(indexes):
        correct_ugrid['order'][ind] = centers_pc['order'][i]
        
plotter = pv.Plotter(notebook=False)

plotter.add_mesh(centers_pc)
plotter.add_mesh(correct_ugrid,style = 'wireframe')
plotter.add_camera_orientation_widget()
plotter.show()

```


![image](https://user-images.githubusercontent.com/40857804/174857206-da91019b-bce5-4fe9-a59d-c740699eb114.png)

The wrong uniform grid on the other hand:
```python
# Plot wrong_ugrid and centers

wrong_ugrid = correct_ugrid.copy()
indexes = wrong_ugrid.find_containing_cell(centers_pc.points)
wrong_ugrid['order'] = centers_pc['order']

plotter = pv.Plotter(notebook=False)

plotter.add_mesh(centers_pc)
plotter.add_mesh(wrong_ugrid,style = 'wireframe')
plotter.add_camera_orientation_widget()
plotter.show()
```
![image](https://user-images.githubusercontent.com/40857804/174857309-6876f076-9169-486a-b613-8933d5f0f4e5.png)


This is a solution but is very slow for high resolution grids. Is the index order switch intentional? Is there a better way to do this?

**Comment by kpan2** [comment]
I would like to know how the large example volumes were built too. I'm also building 3D volumes with UniformGrid but it takes way too long for the tissues I'm constructing. Hopefully this bumps the question somehow!

**Reply by akaszynski** [reply]
Please open a new discussion and (if at all possible) add the dataset in question. This way we'll be able to better help you.

**Reply by kpan2** [reply]
Thanks for the heads up! I have just done so #4373
